### ISSUE TITLE
Apple Metal

### ISSUE THREAD

**Issue by haolange** [dotnet/Silk.NET#1121]
Hi, Is there a plan for suppor metal？

**Comment by Perksey** [comment]
Yes there is a plan.

**Comment by IsaacMarovitz** [comment]
@Perksey Any update on this? Projects like Ryujinx could very much benefit from Metal bindings in the future.

**Comment by haolange** [comment]
Guys, I build a dotnet binding for Metal 3.0 here is the link : https://github.com/haolange/SharpMetal/tree/main

**Comment by IsaacMarovitz** [comment]
> Guys, I build a dotnet binding for Metal 3.0, Here is the link : https://github.com/haolange/SharpMetal/tree/main

Hi there, you seem to have plagiarised [my repository](https://github.com/IsaacMarovitz/SharpMetal). You are within your rights to do so under the MIT license that that code was initially distributed under, but it's generally not a good idea to then advertise that plagiarism on a thread with the original dev :)

**Comment by haolange** [comment]
> > Guys, I build a dotnet binding for Metal 3.0, Here is the link : https://github.com/haolange/SharpMetal/tree/main
> 
> Hi there, you seem to have plagiarised [my repository](https://github.com/IsaacMarovitz/SharpMetal). You are within your rights to do so under the MIT license that that code was initially distributed under, but it's generally not a good idea to then advertise that plagiarism on a thread with the original dev :)

I'm really sorry, this library was given to me by a friend. After asking him, he did indeed download it from your library. I have made some modifications to this library, mainly focusing on iOS support and GC free objects, as well as the content of MetalFX. May I add your library link as the source library for this library? Or should i delete it?

**Comment by IsaacMarovitz** [comment]
> I'm really sorry, this library was given to me by a friend. After asking him, he did indeed download it from your library. I have made some modifications to this library, mainly focusing on iOS support and GC free objects, as well as the content of MetalFX. May I add your library link as the source library for this library? Or should i delete it?

I'm just fine if you link it. Also, consider PRing the GC stuff, as that is a core issue with the current implementation :)

**Comment by haolange** [comment]
> > I'm really sorry, this library was given to me by a friend. After asking him, he did indeed download it from your library. I have made some modifications to this library, mainly focusing on iOS support and GC free objects, as well as the content of MetalFX. May I add your library link as the source library for this library? Or should i delete it?
> 
> I'm just fine if you link it. Also, consider PRing the GC stuff, as that is a core issue with the current implementation :)

Ok i‘m already link it. Thank you and very very sorry. Finally what is PRing GC stuff？

**Comment by PaperPrototype** [comment]
Any progress on this? What about using SharpMetal as a foundation?

**Comment by Perksey** [comment]
The SharpMetal team and the Silk.NET team both agree that wouldn’t be sensible, as their generator technology isn’t completely bulletproof and is only a stopgap solution until something better comes along.

**Comment by IsaacMarovitz** [comment]
> Any progress on this? What about using SharpMetal as a foundation?

The best solution to me is waiting for .NET-Swift interop to land, in which case this should be trivial to solve. But yeah as Perksey the SharpMetal generator is too jank to be used for Silk.

**Comment by PaperPrototype** [comment]
O.o looks like a lot of work has already gone into support for Swift interop https://github.com/dotnet/runtime/issues/93631

**Comment by Perksey** [comment]
Yep the runtime support is all there, we just need to write our generator. The .NET team are working on their own generator but it doesn’t seem to be a high priority work item (and also we write our own generators anyway)

This is something I really really want to work on but I’m bogged down on much less exciting things because we just don’t have the contributors atm.

**Comment by IsaacMarovitz** [comment]
I can take a look a migrating SharpMetal and see how it goes/if I can make an actual stable generator

**Comment by IsaacMarovitz** [comment]
Last time I tried using .NET teams generator it quickly ran into problems

**Comment by Perksey** [comment]
There’s a lot of intricate work involved I think to add this into SilkTouch, so I wouldn’t necessarily wish that upon a contributor. However, a good route of inquiry would be to see if a custom generator could be made to read the SwiftInterface stuff and generate bindings similar to what ClangSharp outputs for its COM bindings (i.e. raw pointers, protocols/interfaces represented as structs, etc) We could then integrate this into SilkTouch the same way we do ClangSharp, and should be achievable for a contributor that doesn’t have an encyclopaedic knowledge of SilkTouch (this is a problem we will solve eventually, we just haven’t got around to writing for-contributors docs yet)

**Comment by Perksey** [comment]
And for the avoidance of doubt, we can reuse e.g. parsing code from the .NET generator - it’s all .NET Foundation code after all.

**Comment by PaperPrototype** [comment]
Checking back after awhile. What is the current status on this? Has anyone managed to make use of SharpMetal?

**Comment by Perksey** [comment]
No plans from our side. If a community contribution wanted to do it we’d accept it but it’s an extreme amount of non-trivial work.

**Comment by PaperPrototype** [comment]
just putting these here 
https://forums.swift.org/t/are-there-any-plans-we-get-swift-dotnet-anytime-soon/80875/3
https://github.com/dotnet/runtimelab/tree/feature/swift-bindings
